Sql Server Fails to Start with DistSQL Error

**Describe the problem**
SQL servers identify themselves in the system.sql_instances table. The table
contains the sql address and rpc address of the sql server. Here are example 
contents of the sql_instances table:

```
test@fresh-tenant-ver-5005.6h5c.crdb.io:26257/defaultdb> select * from system.sql_instances;
  id |                      addr                      |                session_id                |              locality               |                    sql_addr                    | crdb_region | binary_version
-----+------------------------------------------------+------------------------------------------+-------------------------------------+------------------------------------------------+-------------+-----------------
   1 | 10-0-3-139.us-central1.pod.cluster.local:26257 | \x02c1a251a76346fc91267ee61c139e38       | {"Tiers": "region=gcp-us-central1"} | NULL                                           | \x80        | NULL
   2 | 10-0-5-139.us-central1.pod.cluster.local:26257 | \x010180a704ae5254ce4c048c96b79f2f930dfb | {"Tiers": "region=gcp-us-central1"} | 10-0-5-139.us-central1.pod.cluster.local:26257 | \x80        | 23.1

```

If the sql_instance table contains a stale row with an ipaddress+port that is
reused by something else, then sql servers will fail to start with a distsql
error. It looks like a query run during sql server start up attempts to
schedule a dist sql flow on the stale instance and the distsql query failing
causes the process to crash.

This was discovered while attempting to deflake
https://github.com/cockroachdb/cockroach/issues/105402. The bug could happen in
production if an ip address was reused by another tenant or if a sql server
container crashed in a pod and was restarted.

**To Reproduce**
[PR 106538](https://github.com/cockroachdb/cockroach/pull/106538) contains a regression test. Running the test will fail with the following stack trace:

```
--- FAIL: TestStartTenantWithStaleInstance (9.74s)
    test_server_shim.go:467: migration-job-find-already-completed: failed to connect to n1 at 127.0.0.1:46163: initial connection heartbeat failed: grpc: connection error: desc = "transport: authentication handshake failed: context deadline exceeded" [code 14/Unavailable]
        (1) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal.func1.1
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:1032
          | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next.func1
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:477
          | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:528
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternalBuffered
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:642
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:694
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:680
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:664
          | github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable.CheckIfMigrationCompleted
          |     github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable/migrations_table.go:118
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades.func2
          |     github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:220
          | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
          |     github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades
          |     github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:216
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart
          |     github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1627
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).PreStart
          |     github.com/cockroachdb/cockroach/pkg/server/tenant.go:746
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).Start
          |     github.com/cockroachdb/cockroach/pkg/server/tenant.go:913
          | github.com/cockroachdb/cockroach/pkg/server.(*TestServer).StartTenant
          |     github.com/cockroachdb/cockroach/pkg/server/testserver.go:1188
          | github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartTenant
          |     github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:465
          | github.com/cockroachdb/cockroach/pkg/ccl/serverccl.TestStartTenantWithStaleInstance
          |     github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:463
          | [...repeated from below...]
        Wraps: (2) migration-job-find-already-completed
        Wraps: (3) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
          |     github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:181
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).Dial
          |     github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:103
          | github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run
          |     github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:112
          | github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1
          |     github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:152
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
          |     github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484
          | runtime.goexit
          |     GOROOT/src/runtime/asm_amd64.s:1594
        Wraps: (4) failed to connect to n1 at 127.0.0.1:46163
        Wraps: (5) forced error mark
          | "originated at breaker breaker"
          | github.com/cockroachdb/cockroach/pkg/util/circuit/*circuit.breakerErrorMark::
        Wraps: (6) forced error mark
          | "breaker open"
          | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        Wraps: (7) initial connection heartbeat failed
        Wraps: (8) grpc: connection error: desc = "transport: authentication handshake failed: context deadline exceeded" [code 14/Unavailable]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *markers.withMark (6) *markers.withMark (7) *netutil.InitialHeartbeatFailedError (8) *status.Error
```

Jira issue: CRDB-29607

--- a/pkg/ccl/serverccl/BUILD.bazel
+++ b/pkg/ccl/serverccl/BUILD.bazel
@@ -87,6 +87,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/protoutil",
         "//pkg/util/randutil",
+        "//pkg/util/stop",
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_cockroachdb_errors//:errors",

--- a/pkg/ccl/serverccl/server_sql_test.go
+++ b/pkg/ccl/serverccl/server_sql_test.go
@@ -12,6 +12,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"net"
 	"net/http"
 	"strings"
 	"testing"
@@ -35,6 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/errors"
 	"github.com/lib/pq"
 	"github.com/stretchr/testify/require"
@@ -412,4 +414,60 @@ func TestTenantInstanceIDReclaimLoop(t *testing.T) {
 func TestSystemConfigWatcherCache(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	systemconfigwatchertest.TestSystemConfigWatcher(t, false /* skipSecondary */)
+}
+
+// TestStartTenantWithStaleInstance covers the following scenario:
+// - a sql server starts up and is assigned port 'a'
+// - the sql server shuts down and releases port 'a'
+// - something else starts up and claims port 'a'. In the test that is the
+// listener. This is important because the listener causes connections to 'a' to
+// hang instead of responding with a RESET packet.
+// - a different server with stale instance information schedules a distsql
+// flow and attempts to dial port 'a'.
+func TestStartTenantWithStaleInstance(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s := serverutils.StartServerOnly(t, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+	})
+	defer s.Stopper().Stop(ctx)
+
+	var listener net.Listener
+	// In rare cases under stress net.Listen call can result in an error that
+	// the address is already in use (because the stopped tenant hasn't released
+	// the socket); thus, we allow for some retries to go around that issue.
+	testutils.SucceedsSoon(t, func() error {
+		rpcAddr := func() string {
+			tenantStopper := stop.NewStopper()
+			defer tenantStopper.Stop(ctx)
+			server, db := serverutils.StartTenant(t, s, base.TestTenantArgs{
+				Stopper:  tenantStopper,
+				TenantID: serverutils.TestTenantID(),
+			},
+			)
+			defer db.Close()
+			return server.RPCAddr()
+		}()
+
+		var err error
+		listener, err = net.Listen("tcp", rpcAddr)
+		return err
+	})
+	defer func() {
+		_ = listener.Close()
+	}()
+
+	_, db := serverutils.StartTenant(t, s, base.TestTenantArgs{
+		TenantID: serverutils.TestTenantID(),
+	})
+	defer func() {
+		_ = db.Close()
+	}()
+
+	// Query a table to make sure the tenant is healthy, doesn't really matter
+	// which table.
+	_, err := db.Exec("SELECT count(*) FROM system.sqlliveness")
+	require.NoError(t, err)
 }

--- a/pkg/jobs/joberror/BUILD.bazel
+++ b/pkg/jobs/joberror/BUILD.bazel
@@ -8,6 +8,7 @@ go_library(
     deps = [
         "//pkg/kv/kvclient/kvcoord",
         "//pkg/sql/flowinfra",
+        "//pkg/sql/sqlerrors",
         "//pkg/util/circuit",
         "//pkg/util/grpcutil",
         "//pkg/util/sysutil",

--- a/pkg/jobs/joberror/errors.go
+++ b/pkg/jobs/joberror/errors.go
@@ -11,35 +11,15 @@
 package joberror
 
 import (
-	"strings"
-
 	circuitbreaker "github.com/cockroachdb/circuitbreaker"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/sql/flowinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
 	"github.com/cockroachdb/cockroach/pkg/util/circuit"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
 	"github.com/cockroachdb/cockroach/pkg/util/sysutil"
 	"github.com/cockroachdb/errors"
 )
-
-// IsDistSQLRetryableError returns true if the supplied error, or any of its parent
-// causes is an rpc error.
-// This is an unfortunate implementation that should be looking for a more
-// specific error.
-func IsDistSQLRetryableError(err error) bool {
-	if err == nil {
-		return false
-	}
-
-	// TODO(knz): this is a bad implementation. Make it go away
-	// by avoiding string comparisons.
-
-	errStr := err.Error()
-	// When a crdb node dies, any DistSQL flows with processors scheduled on
-	// it get an error with "rpc error" in the message from the call to
-	// `(*DistSQLPlanner).Run`.
-	return strings.Contains(errStr, `rpc error`)
-}
 
 // isBreakerOpenError returns true if err is a circuit.ErrBreakerOpen.
 //
@@ -57,7 +37,7 @@ func IsPermanentBulkJobError(err error) bool {
 	if err == nil {
 		return false
 	}
-	return !IsDistSQLRetryableError(err) &&
+	return !sqlerrors.IsDistSQLRetryableError(err) &&
 		!grpcutil.IsClosedConnection(err) &&
 		!flowinfra.IsFlowRetryableError(err) &&
 		!flowinfra.IsNoInboundStreamConnectionError(err) &&

--- a/pkg/server/drain.go
+++ b/pkg/server/drain.go
@@ -447,10 +447,6 @@ func (s *drainServer) drainClients(
 	// tasks that may issue SQL statements have shut down.
 	s.sqlServer.leaseMgr.SetDraining(ctx, true /* drain */, reporter)
 
-	// Mark this phase in the logs to clarify the context of any subsequent
-	// errors/warnings, if any.
-	log.Infof(ctx, "SQL server drained successfully; SQL queries cannot execute any more")
-
 	session, err := s.sqlServer.sqlLivenessProvider.Release(ctx)
 	if err != nil {
 		return err
@@ -464,8 +460,9 @@ func (s *drainServer) drainClients(
 
 	// Mark the node as fully drained.
 	s.sqlServer.gracefulDrainComplete.Set(true)
-
-	// Done. This executes the defers set above to drain SQL leases.
+	// Mark this phase in the logs to clarify the context of any subsequent
+	// errors/warnings, if any.
+	log.Infof(ctx, "SQL server drained successfully; SQL queries cannot execute any more")
 	return nil
 }
 

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -44,6 +44,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
@@ -1998,7 +1999,8 @@ func (dsp *DistSQLPlanner) PlanAndRun(
 			// cancellation has already occurred.
 			return
 		}
-		if !pgerror.IsSQLRetryableError(distributedErr) &&
+		if !sqlerrors.IsDistSQLRetryableError(distributedErr) &&
+			!pgerror.IsSQLRetryableError(distributedErr) &&
 			!flowinfra.IsFlowRetryableError(distributedErr) &&
 			!isDialErr(distributedErr) {
 			// Only re-run the query if we think there is a high chance of a

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -100,6 +100,22 @@ type runnerResult struct {
 	err    error
 }
 
+type runnerDialErr struct {
+	err error
+}
+
+func (e *runnerDialErr) Error() string {
+	return e.err.Error()
+}
+
+func (e *runnerDialErr) Cause() error {
+	return e.err
+}
+
+func isDialErr(err error) bool {
+	return errors.HasType(err, (*runnerDialErr)(nil))
+}
+
 // run executes the request. An error, if encountered, is both sent on the
 // result channel and returned.
 func (req runnerRequest) run() error {
@@ -111,6 +127,9 @@ func (req runnerRequest) run() error {
 
 	conn, err := req.podNodeDialer.Dial(req.ctx, roachpb.NodeID(req.sqlInstanceID), rpc.DefaultClass)
 	if err != nil {
+		// Mark this error as special runnerDialErr so that we could retry this
+		// distributed query as local.
+		err = &runnerDialErr{err: err}
 		res.err = err
 		return err
 	}
@@ -1979,7 +1998,9 @@ func (dsp *DistSQLPlanner) PlanAndRun(
 			// cancellation has already occurred.
 			return
 		}
-		if !pgerror.IsSQLRetryableError(distributedErr) && !flowinfra.IsFlowRetryableError(distributedErr) {
+		if !pgerror.IsSQLRetryableError(distributedErr) &&
+			!flowinfra.IsFlowRetryableError(distributedErr) &&
+			!isDialErr(distributedErr) {
 			// Only re-run the query if we think there is a high chance of a
 			// successful local execution.
 			return

--- a/pkg/sql/sqlerrors/errors.go
+++ b/pkg/sql/sqlerrors/errors.go
@@ -12,6 +12,8 @@
 package sqlerrors
 
 import (
+	"strings"
+
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
@@ -410,4 +412,23 @@ func errHasCode(err error, code ...pgcode.Code) bool {
 		}
 	}
 	return false
+}
+
+// IsDistSQLRetryableError returns true if the supplied error, or any of its parent
+// causes is an rpc error.
+// This is an unfortunate implementation that should be looking for a more
+// specific error.
+func IsDistSQLRetryableError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// TODO(knz): this is a bad implementation. Make it go away
+	// by avoiding string comparisons.
+
+	errStr := err.Error()
+	// When a crdb node dies, any DistSQL flows with processors scheduled on
+	// it get an error with "rpc error" in the message from the call to
+	// `(*DistSQLPlanner).Run`.
+	return strings.Contains(errStr, `rpc error`)
 }

